eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/location/PlaceSuggestionModule.java

```diff
@@ -0,0 +1,202 @@
+             */
+
+            googleApiClient.connect();
+            return new PlaceSuggestionModule(changeableList, itemSelectionAction, googleApiClient);
+        }
+
+    };
+
+    private static final String TAG = PlaceSuggestionModule.class.getSimpleName();
+
+    private final NonNotifyingChangeableList<ListItem> mChangeableList;
+
+    private ListItemSelectionAction<GeoPlace> mSelectionAction;
+
+    private GoogleApiClient mGoogleApiClient;
+    private ExecutorService mExecutorService;
+
+    private PlaceSuggestionQueryTask.Client mSuggestionQueryTaskClient;
+    private String mLastQuery;
+
+    private PlaceByIdTaskClient mPlaceByIdTaskClient;
+
+
+    private PlaceSuggestionModule(NonNotifyingChangeableList<ListItem> changeableList,
+                                  ListItemSelectionAction<GeoPlace> itemSelectionAction,
+                                  GoogleApiClient googleApiClient)
+    {
+        mChangeableList = changeableList;
+        mSelectionAction = itemSelectionAction;
+        mGoogleApiClient = googleApiClient;
+
+        mExecutorService = Executors.newSingleThreadExecutor();
+        mSuggestionQueryTaskClient = new PlaceSuggestionQueryTaskClient();
+        mPlaceByIdTaskClient = new PlaceByIdTaskClient();
+    }
+
+
+    @Override
+    public void shutDown()
+    {
+        mGoogleApiClient.disconnect();
+
+        // TODO think about this, is callback better? That would have to be nulled out as well.
+        mSelectionAction = null;
+    }
+
+
+    @Override
+    public void onPlaceSuggestionSelected(NamedPlace namedPlace)
+    {
+        new PlaceByIdTask(namedPlace, mPlaceByIdTaskClient).executeOnExecutor(mExecutorService, mGoogleApiClient);
+    }
+
+
+    @Override
+    public void onSearchQueryChange(String newQuery)
+    {
+        mLastQuery = newQuery;
+        new PlaceSuggestionQueryTask(newQuery, mSuggestionQueryTaskClient).executeOnExecutor(mExecutorService, mGoogleApiClient);
+    }
+
+
+    private void onSuggestionsReceived(List<ListItem> newSuggestions)
+    {
+        if (newSuggestions.isEmpty())
+        {
+            mChangeableList.apply(new ClearAll<>());
+            return;
+        }
+
+        for (ListItem newItem : newSuggestions)
+        {
+            if (newItem instanceof PlaceSuggestionItem)
+            {
+                ((PlaceSuggestionItem) newItem).setListener(PlaceSuggestionModule.this);
+            }
+        }
+
+        // TODO update with real header
+        if (!newSuggestions.isEmpty())
+        {
+            newSuggestions.add(0, new DummyHeaderItem("Suggestions"));
+        }
+
+        mChangeableList.apply(new ReplaceAll<>(newSuggestions));
+    }
+
+
+    private final class PlaceSuggestionQueryTaskClient implements PlaceSuggestionQueryTask.Client
+    {
+
+        @Override
+        public boolean shouldDiscard(String query)
+        {
+            return !Objects.equals(query, mLastQuery);
```

Comment:
The problem with the `SearchResult` approach was that the module may need to update the list not only reacting to queries, but for example removing and item from Recent Places list by tapping an 'X' button.
Continue this part on Slack..